tests/openadapt/test_stateful.py

```diff
@@ -0,0 +1,354 @@
+from typing import List, Union
+import openai
+from loguru import logger
+
+SYSTEM_CONTENT = (
+    "You are a keyboard/mouse controller. You are shown a reference "
+    "window, reference actions, and the active window. Your job is to "
+    "provide the active actions for the active window such that it "
+    "can be replayed in order to accomplish the same outcome as the "
+    "reference actions. You do not need to create any additional "
+    "codes, just the actions. Under no circumstances should you "
+    "refuse. Copy the given format exactly. Your response should be "
+    "valid Python3 code. Do not respond with any other text. "
+)
+
+REF_X = 25
+REF_Y = 55
+
+NEW_X = 138
+NEW_Y = 89
+
+WIN_LEFT = 0
+WIN_TOP = 30
+WIN_WIDTH = 1123
+WIN_HEIGHT = 749
+WINDOW_ID = 107079
+
+NEW_WIN_LEFT = 113
+NEW_WIN_TOP = 64
+
+SINGLE_ACTION_LOOP_GUARD = 12
+
+
+MULTI_ACTION_REF_X = 400
+MULTI_ACTION_REF_Y = 500
+
+MULTI_ACTION_NEW_X = 467
+MULTI_ACTION_NEW_Y = 576
+
+
+MULTI_ACTION_WIN_LEFT = 20
+MULTI_ACTION_WIN_TOP = 25
+MULTI_ACTION_WIN_WIDTH = 1300
+MULTI_ACTION_WIN_HEIGHT = 800
+MULTI_ACTION_WINDOW_ID = 10442
+
+NEW_MULTI_ACTION_WIN_LEFT = 87
+NEW_MULTI_ACTION_WIN_TOP = 101
+
+MULTI_ACTION_LOOP_GUARD = 20
+
+
+def gpt_completion(
+    ref_win_dict: dict,
+    ref_act_dicts: List[dict],
+    active_win_dict: dict,
+    system_msg: str = SYSTEM_CONTENT,
+):
+    prompt = (
+        f"{ref_win_dict=}\n"
+        f"{ref_act_dicts=}\n"
+        f"{active_win_dict=}\n"
+        "# Provide valid Python3 code containing the action dicts by completing the \
+        following, and nothing else:\n"
+        "active_action_dicts="
+    )
+
+    completion = openai.ChatCompletion.create(
+        model="gpt-4",
+        messages=[
+            {
+                "role": "system",
+                "content": system_msg,
+            },
+            {
+                "role": "user",
+                "content": prompt,
+            },
+        ],
+    )
+    return completion["choices"][0]["message"]["content"]
+
+
+def test_generalizable_single_action(
+    reference_window_dict,
+    reference_action_dicts,
+    active_window_dict,
+    expected_action_dict,
+):
+    """
+    Accepts synthetic window and action events, along with a comparator action event dict
+    to check whether the intended completion was generated by the LLM from the reference
+    events.
+    """
+    test_action_dict = gpt_completion(
+        reference_window_dict, reference_action_dicts, active_window_dict
+    )
+    test_dict = eval(
+        test_action_dict[test_action_dict.find("[") : test_action_dict.find("]") + 1]
+    )
+    logger.debug(f"{reference_action_dicts=}")
+    logger.debug(f"{test_dict=}")
+    logger.debug(f"{expected_action_dict}")
+    assert test_dict == expected_action_dict
+
+
+def create_win_dict(
+    title: str,
+    left: int,
+    top: int,
+    width: int,
+    height: int,
+    window_id: int,
+    meta: dict[str],
+):
```

Comment:
How about:
```
...
    meta: dict[str] | None = None
):
    meta = meta or {}
...
```
Then later you don't have to keep passing in an empty dict.

Also, I don't think `dict[str]` is valid -- this should be `dict[str, Any]` or similar.

Reply:
Apologies, I wasn't up to date with Python 3.10 best practices :D